Taking Turns

kelseyroy opened the pull request:
Adds:
- Validate class to assess whether a players input is valid and the chosen spot is not occupied
- Validate test suite
- Recursion in select_spot based on what validate_input returns
- take_turns method in TicTacToe
- TicTacToe instance variables for current_player and num_of_turns

Help needed:
- I'm not sure I want GamePlay to have knowledge about UserInterface and Messages, but I'm also not sure how I would refactor to fix this. 
- Is take_turns impure? Could I produce side effects by having these mutable variables? Am I applying functional programming logic where it shouldn't be?

kelseyroy committed:
Validates the players input and raises descriptive errors if the input is invalid

kelseyroy committed:
Select spot shares a helpful message if user submits an invalid spot and allows the player to try again

kelseyroy committed:
TicTacToe can now take turns 9 times

kelseyroy commented on lib/tic_tac_toe.rb:
```diff
@@ -4,17 +4,31 @@
     @ui.display_message(@message.lookup(:welcome))
     @ui.display_message(@message.lookup(:instructions))
     @ui.display_board(@board)
+    while @num_of_turns < 9 do
```
As a heads up, this is sort of a place holder while I build out the game ending logic. I don't expect every game will have exactly 9 turns.

t-keazirian commented on lib/validate.rb:
```diff
@@ -0,0 +1,28 @@
+require_relative 'messages'
+require_relative 'ui'
```
Nice! I like this Validator class.  Looks like you can remove these dependencies here and in the `initialize`!

MarieRobinson replied:
I agree, this is nice!

t-keazirian commented on spec/validate_spec.rb:
```diff
@@ -0,0 +1,22 @@
+          expect(@test_validate.validate_input(2, @board)).to eq(true)
+      end
+
+end
```
You have a few new line at the EOF alerts on this PR.

kelseyroy committed:
Adds

- EOF lines
- take_turns becomes a private method

Removes

- Message and UserInterface from Validate

t-keazirian commented on lib/game_play.rb:
```diff
@@ -22,21 +25,15 @@ def mark_board(board, player_mark, spot)
+        is_valid = @validate.validate_input(spot, board)
+        if !is_valid
+            @ui.display_message(@messages.lookup(:invalid_input)) 
+            return select_spot(board)
```
woo recursion!!!

t-keazirian commented on lib/validate.rb:
```diff
@@ -0,0 +1,20 @@
+
+    private
+
+    def valid_input?(input)
```
Thinking about the naming of these methods...

I'm wondering if `valid_input?` could be changed to `valid_integer?` or something along those lines, to be more specific, and `validate_input` could be called `valid_free_spot?` or whatever you think a good name for that would be. `valid_free_spot?` indicates to me that it returns a boolean. open to disagreements or discussion!

kelseyroy replied:
This is a good call out, I like both of these suggestions! Since my original thought was that `validate_input` would throw an error instead of returning false, I wasn't thinking of it as a boolean. `valid_integer` is also more descriptive so I prefer it as well!

kelseyroy committed:
Changes Validate method names to increase clarity
